### PR TITLE
apiv2 container create using specgen

### PR DESCRIPTION
**Pull request by baude**
this uses the specgen structure to create containers rather than the outdated createconfig.  right now, only the apiv2 create is wired up.  eventually the cli will also have to be done.

Signed-off-by: Brent Baude <bbaude@redhat.com>

### CODE REVIEW

File: libpod/container.go [hunk 1]
```diff
@@ -372,11 +372,11 @@ type ContainerConfig struct {
 	StopTimeout uint `json:"stopTimeout,omitempty"`
 	// Time container was created
 	CreatedTime time.Time `json:"createdTime"`
-	// NoCgroups indicates that the container will not create CGroups. It is
+	// CgroupsMode indicates that the container will not create CGroups. It is
```
Comment (larsks) [inline libpod/container.go]: This looks like some over-enthusiastic search-and-replace...I suspect this should still read `NoCgroups`.

Comment (baude) [reply libpod/container.go]: indeed, good catch

File: libpod/define/errors.go [hunk 2]
```diff
@@ -45,6 +45,7 @@ var (
 	// ErrPodFinalized indicates that the pod has already been created and
 	// cannot be modified
 	ErrPodFinalized = errors.New("pod has been finalized")
+
```
Comment (mheon) [inline libpod/define/errors.go]: ?

File: pkg/api/handlers/libpod/containers_create.go [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+		return
+	}
+	ctr, err := sg.MakeContainer(runtime)
+	if err != nil {
+		utils.InternalServerError(w, err)
+		return
+	}
+	response := utils.ContainerCreateResponse{ID: ctr.ID()}
+	utils.WriteJSON(w, http.StatusCreated, response)
+
```
Comment (mheon) [inline pkg/api/handlers/libpod/containers_create.go]: Extra newline

File: pkg/api/handlers/utils/containers.go [hunk 4]
```diff
@@ -119,3 +129,29 @@ func GenerateFilterFuncsFromMap(r *libpod.Runtime, filters map[string][]string)
 	}
 	return filterFuncs, nil
 }
+
+func CreateContainer(ctx context.Context, w http.ResponseWriter, runtime *libpod.Runtime, cc *createconfig.CreateConfig) {
+	var pod *libpod.Pod
+	ctr, err := shared.CreateContainerFromCreateConfig(runtime, cc, ctx, pod)
+	if err != nil {
+		//if strings.Contains(err.Error(), "invalid log driver") {
```
Comment (mheon) [inline pkg/api/handlers/utils/containers.go]: Can we get rid of this?

File: pkg/spec/config_linux.go [hunk 5]
```diff
@@ -32,6 +32,11 @@ func Device(d *configs.Device) spec.LinuxDevice {
 	}
 }
 
+// DevicesFromPath is an exported wrapper for devicesFromPath
+func DevicesFromPath(g *generate.Generator, devicePath string) error {
```
Comment (mheon) [inline pkg/spec/config_linux.go]: Can we just rename devicesFromPath instead? I don't think we need separate exported/not exported versions

File: pkg/spec/spec.go [hunk 6]
```diff
@@ -404,8 +404,13 @@ func (config *CreateConfig) createConfigToOCISpec(runtime *libpod.Runtime, userM
 	return configSpec, nil
 }
 
-func blockAccessToKernelFilesystems(config *CreateConfig, g *generate.Generator) {
-	if !config.Security.Privileged {
+// BlockAccessToKernelFilesystems is an exported wrapper for blockAccessToKernelFilesystems
+func BlockAccessToKernelFilesystems(privileged, pidModeIsHost bool, g *generate.Generator) {
```
Comment (mheon) [inline pkg/spec/spec.go]: Probably don't need a separate exported version, can just rename the other one

File: pkg/specgen/create.go [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+	"github.com/containers/libpod/libpod/define"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+)
+
+// MakeContainer creates a container based on the SpecGenerator
+func (s *SpecGenerator) MakeContainer(rt *libpod.Runtime) (*libpod.Container, error) {
+	var pod *libpod.Pod
+	if err := s.validate(rt); err != nil {
+		return nil, errors.Wrap(err, "failed to create container")
```
Comment (mheon) [inline pkg/specgen/create.go]: Should this be something more like "invalid config provided"?

File: pkg/specgen/create.go [hunk 8]
```diff
@@ -0,0 +1,203 @@
...
+	}
+	options = append(options, s.createExitCommandOption(rtc, podmanPath))
+	newImage, err := rt.ImageRuntime().NewFromLocal(s.Image)
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO mheon wants to talk with Dan about this
+	useImageVolumes := s.ImageVolumeMode == "bind"
+	options = append(options, libpod.WithRootFSFromImage(newImage.ID(), s.Image, useImageVolumes))
```
Comment (mheon) [inline pkg/specgen/create.go]: You know, I think useImageVolumes might be 100% unused...

File: pkg/specgen/create.go [hunk 9]
```diff
@@ -0,0 +1,203 @@
...
+	if len(s.Name) > 0 {
+		logrus.Debugf("setting container name %s", s.Name)
+		options = append(options, libpod.WithName(s.Name))
+	}
+	if s.Pod != "" {
+		logrus.Debugf("adding container to pod %s", s.Pod)
+		options = append(options, rt.WithPod(pod))
+	}
+
+	//if len(mounts) != 0 || len(namedVolumes) != 0 {
```
Comment (mheon) [inline pkg/specgen/create.go]: I think we can still do this with s.Mounts and s.NamedVolumes

Comment (mheon) [reply pkg/specgen/create.go]: Though... Crap, that doesn't include VolumesFrom.

File: pkg/specgen/create.go [hunk 10]
```diff
@@ -0,0 +1,203 @@
...
+	//
+	//if len(namedVolumes) != 0 {
+	//	options = append(options, libpod.WithNamedVolumes(namedVolumes))
+	//}
+	//
+	if len(s.Command) != 0 {
+		options = append(options, libpod.WithCommand(s.Command))
+	}
+
+	options = append(options, libpod.WithEntrypoint(s.Entrypoint))
```
Comment (mheon) [inline pkg/specgen/create.go]: Shouldn't this also be only if len(s.Entrypoint) > 0?

Comment (mheon) [reply pkg/specgen/create.go]: Hmmm. We might want to pull from the image in that case.

File: pkg/specgen/create.go [hunk 11]
```diff
@@ -0,0 +1,203 @@
...
+		options = append(options, libpod.WithStopTimeout(*s.StopTimeout))
+	}
+
+	if s.LogConfiguration != nil {
+		if len(s.LogConfiguration.Path) > 0 {
+			options = append(options, libpod.WithLogPath(s.LogConfiguration.Path))
+		}
+		if len(s.LogConfiguration.Options) > 0 {
+			// Note: I'm really guessing here.
+			options = append(options, libpod.WithLogTag(s.LogConfiguration.Options["tag"]))
```
Comment (mheon) [inline pkg/specgen/create.go]: Should only do if Options["tag"] exists

File: pkg/specgen/create.go [hunk 12]
```diff
@@ -0,0 +1,203 @@
...
+		}
+
+		if len(s.LogConfiguration.Driver) > 0 {
+			options = append(options, libpod.WithLogDriver(s.LogConfiguration.Driver))
+		}
+	}
+
+	// Security options
+
+	options = append(options, libpod.WithSecLabels(s.SelinuxOpts))
```
Comment (mheon) [inline pkg/specgen/create.go]: @rhatdan Should we only do this if explicitly set to something?

File: pkg/specgen/create.go [hunk 13]
```diff
@@ -0,0 +1,203 @@
...
+	// Get namespace related options
+	namespaceOptions, err := s.generateNamespaceContainerOpts(rt)
+	if err != nil {
+		return nil, err
+	}
+	options = append(options, namespaceOptions...)
+
+	// TODO NetworkNS still needs to be done!
+
+	options = append(options, libpod.WithConmonPidFile(s.ConmonPidFile))
```
Comment (mheon) [inline pkg/specgen/create.go]: I think we only do this if ConmonPidFile != ""

File: pkg/specgen/create.go [hunk 14]
```diff
@@ -0,0 +1,203 @@
...
+		options = append(options, libpod.WithRootFS(s.Rootfs))
+	}
+	// Default used if not overridden on command line
+
+	if s.RestartPolicy != "" {
+		if s.RestartPolicy == "unless-stopped" {
+			return nil, errors.Wrapf(define.ErrInvalidArg, "the unless-stopped restart policy is not supported")
+		}
+
+		split := strings.Split(s.RestartPolicy, ":")
```
Comment (mheon) [inline pkg/specgen/create.go]: This is wrong - RestartPolicy is already split in SpecGenerator into RestartPolicy and RestartRetries

File: pkg/specgen/namespaces.go [hunk 15]
```diff
@@ -0,0 +1,404 @@
...
+}
+
+// IsPrivate indicates the namespace is private
+func (n *Namespace) IsPrivate() bool {
+	return n.NSMode == Private
+}
+
+// validate perform simple validation on the namespace to make sure it is not
+// invalid from the get-go
+func (n *Namespace) validate() error {
```
Comment (mheon) [inline pkg/specgen/namespaces.go]: We should check that NSMode is one of the valid ones listed above

File: pkg/specgen/namespaces.go [hunk 16]
```diff
@@ -0,0 +1,404 @@
...
+		}
+	}
+	return nil
+}
+
+func (s *SpecGenerator) generateNamespaceContainerOpts(rt *libpod.Runtime) ([]libpod.CtrCreateOption, error) {
+	options := make([]libpod.CtrCreateOption, 0)
+
+	// Cgroups
+	if s.CgroupNS.IsPrivate() {
```
Comment (mheon) [inline pkg/specgen/namespaces.go]: All of these should probably be switch statements, so we have a default for not-handled modes

File: pkg/specgen/oci.go [hunk 17]
```diff
@@ -0,0 +1,260 @@
...
+
+	isRootless := rootless.IsRootless()
+	if isRootless {
+		inUserNS = true
+	}
+	if !s.UserNS.IsHost() {
+		if s.UserNS.IsContainer() || s.UserNS.IsPath() {
+			inUserNS = true
+		}
+		if s.IDMappings != nil && (len(s.IDMappings.UIDMap) > 0 || len(s.IDMappings.GIDMap) > 0) {
```
Comment (mheon) [inline pkg/specgen/oci.go]: We might be able to set this to `s.UserNS.IsPrivate()` instead

File: pkg/specgen/create.go (hunk 8 above)
Comment (baude) [reply pkg/specgen/create.go]: what do you want me to do; it is in the current code path.

**Comment by mheon** [comment]
I'll look into useImageVolumes on the Libpod side, and make a PR to tear it
out of it really is unused, we can merge that before or after this, doesn't
really matter

File: pkg/api/handlers/libpod/containers_create.go [hunk 18]
```diff
@@ -0,0 +1,30 @@
...
+	"github.com/pkg/errors"
+)
+
+// CreateContainer takes a specgenerator and makes a container. It returns
+// the new container ID on success along with any warnings.
+func CreateContainer(w http.ResponseWriter, r *http.Request) {
+	runtime := r.Context().Value("runtime").(*libpod.Runtime)
+	var sg specgen.SpecGenerator
+	if err := json.NewDecoder(r.Body).Decode(&sg); err != nil {
+		utils.Error(w, "Something went wrong.", http.StatusInternalServerError, errors.Wrap(err, "Decode()"))
```
Comment (TomSweeneyRedHat) [inline pkg/api/handlers/libpod/containers_create.go]: Not a big fan of "Something went wrong".  I'd much prefer just "Error: "

File: pkg/api/handlers/utils/containers.go [hunk 19]
```diff
@@ -119,3 +129,29 @@ func GenerateFilterFuncsFromMap(r *libpod.Runtime, filters map[string][]string)
...
+		//	w.Header().Set("Content-Type", "text/plain; charset=us-ascii")
+		//	w.WriteHeader(http.StatusInternalServerError)
+		//	msg := fmt.Sprintf("logger: no log driver named '%s' is registered", input.HostConfig.LogConfig.Type)
+		//	if _, err := fmt.Fprintln(w, msg); err != nil {
+		//		log.Errorf("%s: %q", msg, err)
+		//	}
+		//	//s.WriteResponse(w, http.StatusInternalServerError, fmt.Sprintf("logger: no log driver named '%s' is registered", input.HostConfig.LogConfig.Type))
+		//	return
+		//}
+		Error(w, "Something went wrong.", http.StatusInternalServerError, errors.Wrap(err, "CreateContainerFromCreateConfig()"))
```
Comment (TomSweeneyRedHat) [inline pkg/api/handlers/utils/containers.go]: Something nit

File: pkg/specgen/config_linux_cgo.go [hunk 20]
```diff
@@ -0,0 +1,62 @@
...
+	}
+
+	if scp == seccomp.PolicyImage {
+		labels, err := img.Labels(context.Background())
+		if err != nil {
+			return nil, err
+		}
+		imagePolicy := labels[seccomp.ContainerImageLabel]
+		if len(imagePolicy) < 1 {
+			return nil, errors.New("no secomppolicy defined by image")
```
Comment (TomSweeneyRedHat) [inline pkg/specgen/config_linux_cgo.go]: nit "no secomppolicy" -> "secomppolicy is not"

File: pkg/specgen/create.go [hunk 21]
```diff
@@ -0,0 +1,203 @@
...
+				return nil, errors.Wrapf(define.ErrInvalidArg, "restart policy requires a positive number of retries")
+			}
+			options = append(options, libpod.WithRestartRetries(uint(numTries)))
+		}
+		options = append(options, libpod.WithRestartPolicy(split[0]))
+	}
+
+	if s.ContainerHealthCheckConfig.HealthConfig != nil {
+		options = append(options, libpod.WithHealthCheck(s.ContainerHealthCheckConfig.HealthConfig))
+		logrus.Debugf("New container has a health check")
```
Comment (TomSweeneyRedHat) [inline pkg/specgen/create.go]: can you include the container name/id here in the message?

Comment (baude) [reply pkg/specgen/create.go]: no  because the container is not created yet

Comment (mheon) [reply pkg/specgen/create.go]: Should we just drop this, then? I don't know if there's value in printing this on the server side if we know nothing more...

**Comment by mheon** [comment]
@giuseppe @rhatdan PTAL - I'd love another set of eyes on this before we merge

File: libpod/runtime_ctr.go [hunk 22]
```diff
@@ -165,10 +165,10 @@ func (r *Runtime) setupContainer(ctx context.Context, ctr *Container) (c *Contai
 		ctr.ociRuntime = ociRuntime
 	}
 
-	// Check NoCgroups support
+	// Check CgroupsMode support
```
Comment (rhatdan) [inline libpod/runtime_ctr.go]: This comment, versus the lines below, does not make sense.

File: pkg/specgen/config_linux_cgo.go (hunk 20 above)
Comment (rhatdan) [inline pkg/specgen/config_linux_cgo.go]: no seccomp policy defined by image

File: pkg/specgen/create.go (hunk 12 above)
Comment (rhatdan) [reply pkg/specgen/create.go]: I would hope the WithSecLabels ignores if nothing is set.

File: libpod/container.go [hunk 23]
```diff
@@ -372,11 +372,11 @@ type ContainerConfig struct {
...
 	// incompatible with CgroupParent.  Deprecated in favor of CgroupsMode.
 	NoCgroups bool `json:"noCgroups,omitempty"`
 	// CgroupsMode indicates how the container will create cgroups
-	// (disabled, no-conmon, enabled).  It supersedes NoCgroups.
+	// (disabled, no-conmon, enabled).  It supersedes CgroupsMode.
```
Comment (mheon) [inline libpod/container.go]: Aggressive rename

File: libpod/container.go (hunk 1 above)
Comment (mheon) [reply libpod/container.go]: Aggressive rename

File: libpod/container_internal.go [hunk 24]
```diff
@@ -1140,7 +1140,7 @@ func (c *Container) stop(timeout uint) error {
 	// the containers cgroup in order to make sure the container is stopped.
 	all := !c.hasNamespace(spec.PIDNamespace)
 	// We can't use --all if CGroups aren't present.
-	// Rootless containers with CGroups v1 and NoCgroups are both cases
+	// Rootless containers with CGroups v1 and CgroupsMode are both cases
```
Comment (mheon) [inline libpod/container_internal.go]: Aggressive rename

File: libpod/options.go [hunk 25]
```diff
@@ -1112,7 +1112,7 @@ func WithCgroupParent(parent string) CtrCreateOption {
 		}
 
 		if ctr.config.NoCgroups {
-			return errors.Wrapf(define.ErrInvalidArg, "CgroupParent conflicts with NoCgroups")
+			return errors.Wrapf(define.ErrInvalidArg, "CgroupParent conflicts with CgroupsMode")
```
Comment (mheon) [inline libpod/options.go]: Aggressive Rename

File: libpod/runtime_ctr.go [hunk 26]
```diff
@@ -168,7 +168,7 @@ func (r *Runtime) setupContainer(ctx context.Context, ctr *Container) (c *Contai
 	// Check NoCgroups support
 	if ctr.config.NoCgroups {
 		if !ctr.ociRuntime.SupportsNoCgroups() {
-			return nil, errors.Wrapf(define.ErrInvalidArg, "requested OCI runtime %s is not compatible with NoCgroups", ctr.ociRuntime.Name())
+			return nil, errors.Wrapf(define.ErrInvalidArg, "requested OCI runtime %s is not compatible with CgroupsMode", ctr.ociRuntime.Name())
```
Comment (mheon) [inline libpod/runtime_ctr.go]: Aggressive Rename

File: libpod/runtime_ctr.go [hunk 27]
```diff
@@ -194,7 +194,7 @@ func (r *Runtime) setupContainer(ctx context.Context, ctr *Container) (c *Contai
 	// Otherwise, the OCI runtime won't be able to stop our container.
 	if ctr.config.NoCgroups {
 		if ctr.config.Spec.Linux == nil {
-			return nil, errors.Wrapf(define.ErrInvalidArg, "must provide Linux namespace configuration in OCI spec when using NoCgroups")
+			return nil, errors.Wrapf(define.ErrInvalidArg, "must provide Linux namespace configuration in OCI spec when using CgroupsMode")
```
Comment (mheon) [inline libpod/runtime_ctr.go]: Aggressive rename

**Pull request merged**